### DISCUSSION TITLE
Parameters *always* not covariant?

### DISCUSSION THREAD

**Discussion by robertmaxton42** [microsoft/pyright#5773] [Q&A] [answered]
Say I've defined a Protocol `Ring` that represents a [ring](https://en.wikipedia.org/wiki/Ring_(mathematics)), and now I want to build on it, say by defining a protocol for [(associative) algebras](https://en.wikipedia.org/wiki/Associative_algebra). If I write

```python
R_co = TypeVar("R", bound=Ring, covariant=True)
class Algebra(Ring, Protocol[R_co]):                                # An associative algebra over a ring R is itself a ring
    @type_ext.override                                              # with a scalar multiplication operator 
    def __mul__(self, other: R_co, /)  -> Self: ...                 # that we associate with 'normal' multiplication
    def __rmul__(self, other: R_co, /) -> Self:                     # and define a right-mul for so we can handle 3 * x
        return self.__mul__(other)
    def __matmul__(self, other: Self, /)   -> Self: ...             # The ring product, now the algebra product, has moved here
                                                                    # and is no longer expected to be commutative
```

Then Pyright complains at me, because apparently covariant variables aren't allowed in parameters. But this seems incorrect: If, say, `R` is `float`, I must be able to use an `int` which subtypes `float` safely -- I've essentially already defined `mul` for them implicitly -- but I shouldn't be allowed to put in a `complex` because I might depend on my scalars being real at some point. 

Am I missing something?

**Comment by erictraut** [comment]
It's hard to follow the code above because it references a bunch of symbols that aren't defined. Could you post a minimal, self-contained code sample?

**Reply by robertmaxton42** [reply]
.... Actually, nevermind, I figured out my error. (I was thinking about the function itself as constraining the generic variable, rather than the _Protocol_ as constraining the type of the _function_; so I was off by a parity flip.) Thanks anyway!

**Comment by robertmaxton42** [comment, marked as answer]
Answer: Yes, if you have a function that takes an argument of generic type, that type should be covariant. But if you have a higher-order function (which is basically what `Protocol` is) that takes a _function_ of generic type, then you insert an extra "sign flip" and the resulting parameter should be _contra_variant (it's fine if it can handle more general types than you asked for, but not if it can handle less general types.) Oops.

Put another way, Callable is just universally contravariant in its first argument, and classes are callables.

And indeed, Pyright will correctly allow you to use a covariant type variable in a case like this:

```python
class Foo(Protocol[T_co]):
    def bar(self, other: Callable[[T_co], Any] , /): ...
```
